discovery-provider/src/eth_indexing/event_scanner.py

```diff
@@ -0,0 +1,376 @@
+        web3: Web3,
+        contract: Type[Contract],
+        state: EventScannerState,
+        event_type: Type[ContractEvent],
+        filters: dict,
+    ):
+        """
+        :param web3: Web3 instantiated with provider url
+        :param contract: Contract
+        :param event_type: web3 Event we scan
+        :param filters: Filters passed to get_logs e.g. { "address": <token-address> }
+        """
+
+        self.logger = logger
+        self.contract = contract
+        self.web3 = web3
+        self.state = state
+        self.event_type = event_type
+        self.filters = filters
+
+    def get_block_timestamp(self, block_num) -> Union[datetime.datetime, None]:
+        """Get Ethereum block timestamp"""
+        try:
+            block_info = self.web3.eth.getBlock(block_num)
+        except BlockNotFound:
+            # Block was not mined yet,
+            # minor chain reorganisation?
+            return None
+        last_time = block_info["timestamp"]
+        return datetime.datetime.utcfromtimestamp(last_time)
+
+    def get_suggested_scan_end_block(self):
+        """Get the last mined block on Ethereum chain we are following."""
+
+        # Do not scan all the way to the final block, as this
+        # block might not be mined yet
+        return self.web3.eth.blockNumber - ETH_BLOCK_TAIL_OFFSET
+
+    def get_last_scanned_block(self) -> int:
+        return self.state.get_last_scanned_block()
+
+    def delete_potentially_forked_block_data(self, after_block: int):
+        """Purge old data in the case of blockchain reorganisation."""
+        self.state.delete_potentially_forked_data(after_block)
+
+    def scan_chunk(self, start_block, end_block) -> Tuple[int, datetime.datetime, list]:
+        """Read and process events between to block numbers.
+
+        Dynamically decrease the size of the chunk if the case JSON-RPC server pukes out.
+
+        :return: tuple(actual end block number, when this block was mined, processed events)
+        """
+
+        block_timestamps = {}
+        get_block_timestamp = self.get_block_timestamp
+
+        # Cache block timestamps to reduce some RPC overhead
+        # Real solution might include smarter models around block
+        def get_block_when(block_num):
+            if block_num not in block_timestamps:
+                block_timestamps[block_num] = get_block_timestamp(block_num)
+            return block_timestamps[block_num]
+
+        all_processed = []
+
+        # Callable that takes care of the underlying web3 call
+        def _fetch_events(_start_block, _end_block):
+            return _fetch_events_for_all_contracts(
+                self.web3,
+                self.event_type,
+                self.filters,
+                from_block=_start_block,
+                to_block=_end_block,
+            )
+
+        # Do `n` retries on `eth_get_logs`,
+        # throttle down block range if needed
+        end_block, events = _retry_web3_call(  # type: ignore
```

Comment:
I don't think this can return None, which hopefully means you can remove the ignore type 🤞

Reply:
Unf currently no typing of exception it seems. will have to put the ignore somewhere. changing to co-locate the type: ignore to the func signature of _retry_web3_call for easier read